Blogs filter

Zineb-Ada opened the pull request:
[Work in progress] This PR includes the following updates:

- Updated markdown files with the implementation of web scraping. This scraping process allowed for the addition of relevant tags to the blogs located in the '/data/planet' directory. These tags will enhance the searchability and categorization of the blogs.
- Implemented a filtering bar in the blogs section. This new feature provides users with a convenient way to filter and refine their search based on specific criteria, making it easier to find the desired content.
These changes aim to improve the overall user experience by enhancing the organization and accessibility of the blogs.

gasche commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
````diff
@@ -0,0 +1,877 @@
+
+```ocamltop
+print_newline;;
+```
````
If you have the space, here may be a good place to point out that `print_newline;;` does *not* print anything, `print_newline ();;` does. To some beginners it may not be clear that the computations (and effects) of a function are only "performed" when the function is called, and that it can otherwise be passed around like any inert value.

johnwhitington replied:
Good idea. Done.

gasche reviewed (commented):
Nice! See some minor comments inline.

gasche commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
````diff
@@ -0,0 +1,877 @@
+done
+
+read_line ()
+```
````
Note: as currently written, none of the examples can actually be "tried", they are not valid toplevel programs (no `;;`) and they cannot be compiled either, due to the ambiguous use of expressions.

johnwhitington replied:
Ok, I added in the double semicolons but also tweaked the language so as not to give the impression double semicolons are part of the language.

gasche replied:
Another approach would be to avoid stray expressions at the toplevel, using `let () = foo` (and then either `let _ = read_line ()` or `let () = ignore (read_line ())`, I prefer the latter). You don't need double-semicolons then because parsing is unambiguous, and actually I think that pasting this form in the toplevel (with just one final `;;`) works well.

johnwhitington commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
````diff
@@ -0,0 +1,877 @@
+
+```ocamltop
+let rec factorial x =
+  if x <= 1 then 1 else x * factorial (x - 1);;
````
Fixed.

johnwhitington commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
````diff
@@ -0,0 +1,877 @@
+let rec range a b =
+  if a > b then []
+  else a :: range (a + 1) b;;
+```
````
Done.

johnwhitington commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
````diff
@@ -0,0 +1,877 @@
+a binary tree carrying any kind of data:
+
+```ocamltop
+type 'a tree = Lf | Br of 'a tree * 'a * 'a tree;;
````
Fixed.

johnwhitington commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
````diff
@@ -0,0 +1,877 @@
+```
+
+The other way OCaml deals with exceptional situations is to use the built-in
+polymorphic `option` type, which is defined as:
````
Fixed.

Octachron commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
````diff
@@ -0,0 +1,877 @@
+  (a +. b) /. 2.0;;
+```
+
+This is rather unusual. In other languages integers get promoted to floating
````
Later, you are using C as an example of "other languages", would it make sense to mention it here?

johnwhitington replied:
Thanks. Fixed.

johnwhitington commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
````diff
@@ -0,0 +1,877 @@
+surrounding it with parentheses and spaces, to give the most concise version:
+
+```ocamltop
+map (map (( * ) 2)) [[1; 2]; [3; 4]; [5; 6]];;
````
Ok, I agree. I'll cut this one.

johnwhitington commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
```diff
@@ -0,0 +1,877 @@
+
+Notice that we do not need to explicitly free memory for such data structures
+when we no longer need it: OCaml is a garbage-collected language, and will free
+memory for data structures when they are no longer needed.
```
Ok, done.

Octachron commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
```diff
@@ -0,0 +1,877 @@
+
+Each expression has one and only one type.
+
+OCaml works out types automatically so you will rarely if ever need to
```
The `if ever` modal seems a bit too strong when explicit types are needed for module types, type-directed disambiguation, GADTs, higher rank polymorphism, and polymorphic recursion....

johnwhitington replied:
Fixed.

johnwhitington commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
````diff
@@ -0,0 +1,877 @@
+let list_find_opt p l =
+  try Some (List.find p l) with
+    Not_found -> None;;
+```
````
I didn't want to delete this (because meaningful small exception examples are hard enough to find), so instead I added yours afterward.

johnwhitington commented:
Ok, all fixed. Thanks everyone.

avsm commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
```diff
@@ -0,0 +1,893 @@
+should install `rlwrap` through your system package manager and run `rlwrap
+ocaml` instead to get history navigation.)
+
+The alternative REPL [utop](https://github.com/diml/utop) may be installed
```
```suggestion
The alternative REPL [utop](https://github.com/ocaml-community/utop) may be installed
```

new homepage for utop

avsm commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
```diff
@@ -0,0 +1,893 @@
+ocaml` instead to get history navigation.)
+
+The alternative REPL [utop](https://github.com/diml/utop) may be installed
+through [OPAM](/docs/install.html#OPAM) or your system package manager. It has
```
```suggestion
through [opam](/docs/install.html#OPAM) or your system package manager. It has
```

lowercase preferred for opam

avsm commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
```diff
@@ -0,0 +1,893 @@
+## Expressions
+
+Our phrase `50 * 50` was an expression, which evaluated to `2500`. OCaml told
+us that the type was `int`, an integer. To avoid repetition, we can give a name
```
This is the first mention of "type". It might be worth a sentence explaining that every value has a type in OCaml before using the term.

johnwhitington replied:
Fixed.

avsm commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
````diff
@@ -0,0 +1,893 @@
+```text
+OCaml type  Range
+
+int         63-bit signed int on 64-bit processors, or 31-bit signed int on
````
a subtlety perhaps, but it can also be 32-bits on js_of_ocaml. Perhaps note that this is for native code generation, but that might be too much info here.

johnwhitington replied:
Yes, I think probably too much info.

avsm commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
````diff
@@ -0,0 +1,893 @@
+```ocamltop
+let f a b =
+  if b = 0 then None else Some (a / b)
+```
````
It's a little odd not to mention type `result` here, which is the preferred way of not "gobbling" errors by returning a None with no further information.

johnwhitington replied:
I just looked at the Stdlib documentation for Result, and it looks like we need to improve it -- there's really no explanation of the expected way to use the functions.

Do you think introducing just the result type without introducing the monadic stuff (which I don't want to do) is worth it?

avsm replied:
It's difficult to use `Result.t` without pointing to one of the libraries that consume the type:
- Rresult: https://github.com/dbuenzli/rresult/blob/master/src/rresult.mli
- Core: http://dev.realworldocaml.org/error-handling.html#encoding-errors-with-result

It's probably worth deferring this to the "second hour with OCaml" indeed

johnwhitington replied:
Ok. By the way, what do you think of this blog post?

https://keleshev.com/composable-error-handling-in-ocaml

I wonder if we might persuade him to let us reuse this content in another of the tutorials...

gasche replied:
A minor note: this style (result monad with polymorphic variants for errors) was advertized in the [Catch me if you can](https://hal.inria.fr/inria-00432575) paper of Teller, Spiwack and Varoquaux back in 2008.

(My suspicion is that it didn't catch on because it is a bit too complex to convince programmers in simple ecosystems. But maybe the lack of traction was also due to the tools of the time (no opam, no standard "result" type, a camlp4 syntax extension instead of binding operators...), and it could have more success today.)

dbuenzli replied:
A minor note on the minor note :-)  

This was actually discussed before the paper on the caml-list mailing list [here](https://inbox.ocaml.org/caml-list/F5231A73-DA8B-4FE8-8B7F-C478F1A65585@erratique.ch/). The page for the mentioned OSR is archived [here](https://web.archive.org/web/20090109050841/http:/cocan.org/osr/exceptionless_error_management).

avsm replied:
Sounds like a separate section on error handling in a future PR is appropriate with all these rich references. There's also a mini-guide to using errors with polymorphic variants over at https://mirage.io/wiki/mirage-3.0-errors.  But for now, this particular PR comment is resolved.

dbuenzli replied:
FWIW there's also `rresult` [guidelines](https://erratique.ch/software/rresult/doc/Rresult.html#usage).

avsm commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
```diff
@@ -0,0 +1,893 @@
+OCaml comes with a library of useful modules which are available anywhere OCaml
+is. For example there are standard libraries for functional data structures
+(such as maps and sets) and imperative data structures (such as hash tables),
+and for making system calls. We use them by writing the module, followed by a
```
```suggestion
and for interacting with the operating system. We use them by writing the module, followed by a
```

`Unix` is not quite as raw as making syscalls directly :-)

avsm commented on site/learn/tutorials/a_first_hour_with_ocaml.md:
```diff
@@ -0,0 +1,893 @@
+## A module from OPAM
+
+Apart from the standard library, a much wider range of modules are available
+through the OCaml Package Manager, OPAM. You must have OCaml on your computer
```
```suggestion
through the OCaml Package Manager, opam. You must have OCaml on your computer
```

avsm reviewed (commented):
Really nice work, John! Comments left in the review.

avsm reviewed (approved):
Looks good to merge. @johnwhitington, would you be able to hook up the indexes first?

johnwhitington commented:
Thanks @avsm 

Done, although I haven't tested it yet, because merging master in broke my build:

```
File "http.ml", line 41, characters 4-23:
41 |   | `Permanent_redirect
         ^^^^^^^^^^^^^^^^^^^
Error: This pattern matches values of type [? `Permanent_redirect ]
       but a pattern was expected which matches values of type
         Cohttp__.Code.status_code
       The second variant type does not allow tag(s) `Permanent_redirect
make[1]: *** [script/rss2html] Error 2
make: *** [local] Error 2
```

I suppose I need to `make deps` again, or re-install something via opam? Any idea?

avsm commented:
Yes you need to `opam update` to pull in the latest version of cohttp; I merged in some changes to support cohttp 4.0 in #1309

avsm commented:
Also need to update the tutorial list over at https://staging.ocaml.org/learn/ to include the new content

johnwhitington commented:
Ah yes. Coming up...

johnwhitington commented:
> Yes you need to `opam update` to pull in the latest version of cohttp; I merged in some changes to support cohttp 4.0 in #1309

I updated and upgraded successfully with opam, but now I get:

```
File "code.ml", line 142, characters 7-28:
142 |     | [Omd.Html_block(_,_,o)] -> o
             ^^^^^^^^^^^^^^^^^^^^^
Error: This pattern matches values of type Omd.block_desc
       but a pattern was expected which matches values of type Omd.block
make[1]: *** [script/md_preprocess] Error 2
make: *** [local] Error 2
```

johnwhitington commented:
Ok, I fixed up the main learn page. I'm not sure about the selection chosen for the front page generally, but I'll look at that separately.

johnwhitington commented:
It looks like the issue above is due to a very recent version of omd where the AST has changed...

```
omd                     2.0.0~alpha1 A Markdown frontend in pure OCaml
```

avsm commented:
That's right -- the constraints on the `ocamlorg.opam` were updated: https://github.com/ocaml/ocaml.org/blob/master/ocamlorg.opam#L12

It would be quite nice if an opam workflow noticed that the constraints changed and had some way to prompt the user that the current packages no longer satisfied the build needs without an `opam update`. I've made a note to bring this up in the opam 2.2 design discussions.

avsm commented:
Verified in staging, and now live. Thanks @johnwhitington and all the reviewers!

Zineb-Ada committed:
Add filtring to Ocaml blogs

Zineb-Ada committed:
Add filtring to Ocaml blogs

Zineb-Ada committed:
Add filtring to Ocaml blogs

SaySayo committed:
added difficulty tags to ocaml books (#1404)

theteachr committed:
Fix grammar error (#1407) (#1408)

sabine committed:
Remove .xml paths from scraper workflow (#1409)

github-actions[bot] committed:
Update scraped data (#1410)

AryanGodara committed:
Contribute a blog post link about Setting up Ocaml and fixing the dependency install error using SymLinks, on MacOS M1/M2 (#1426)

sabine committed:
resolve style conflict on links / tailwind-typography vs styles.css (#1428)

sabine committed:
Tutorials Grouped By Section + Best Practices Broken Up (#1429)

* Dune capitalisation and formatting
Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

---------

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

sabine committed:
Package docs: Read README/LICENSE/CHANGES from status.json (#1435)

* adjust parsing of module map and documentation status to match new voodoo output
* point to docs-data / current

sabine committed:
point to docs-data.ocaml.org/live (#1437)

sabine committed:
in case documentation of latest pkg version is missing, redirect (#1438)

sabine committed:
convert evaluation example from LaTeX to markdown (#1439)

Shreyas-21 committed:
update link checking workflow (#1436)

sabine committed:
use voodoo_serialize to decode status.json and package.json (#1442)

sabine committed:
add curl dependency on CI action (#1444)

sabine committed:
Regular curl in CI (#1445)

sabine committed:
point to docs-data.ocaml.org/live (#1443)

sabine committed:
CI.yml: libcurl4-gnutls-dev (#1446)

sabine committed:
CI: Install libcurl4-openssl-dev in ubuntu-latest runner (#1447)

Turns out the `curl-config` failure is related to these:
- https://github.com/actions/runner-images/issues/7958
- https://github.com/actions/runner-images/issues/7945

Explicitly installing curl on macOS shouldn't be necessary, while on `ubuntu-latest`, we have to install `libcurl4-openssl-dev`.

sabine committed:
CI.yml: run `sudo apt update` before installing libcurl (#1449)

AryanGodara committed:
Add a blog post link @AryanGodara Outreachy internship (#1448)

sabine committed:
install curl in scrape.yml (#1451)

github-actions[bot] committed:
[create-pull-request] automated change (#1453)

sabine committed:
Linkcheck.yml: don't check the HTML bodies from data/planet (#1454)

sabine committed:
Playground: make gen_stdlib.sh POSIX-compatible (#1456)

emillon committed:
Add changelog for dune.3.9.2 (#1458)

* Add changelog for dune.3.9.2

* Only one capitalisation edit

---------

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

github-actions[bot] committed:
[create-pull-request] automated change (#1460)

Co-authored-by: cuihtlauac <cuihtlauac@users.noreply.github.com>

rjbou committed:
[changelog][opam] 2023-07-26-opam-2-2-0-alpha2.md (#1457)

Zineb-Ada committed:
Merge branch 'blogs_filter' of github.com:Zineb-Ada/ocaml.org into blogs_filter

Zineb-Ada committed:
Add tags on data/planet-sources.yml

Zineb-Ada committed:
Add tags on data/planet-sources.yml

Zineb-Ada committed:
Add the tags corresponding to the markdown files in data/planet/-*/*.md

Zineb-Ada committed:
Add the tags corresponding to the markdown files in data/planet/-*/*.md

Zineb-Ada committed:
Deleted the old tags (filtered by institution) and targeted the new tag that I added in `data/planet-source.yml`

Zineb-Ada committed:
Deleted the old tags (filtered by institution) and targeted the new tag that I added in `data/planet-source.yml`

just-max committed:
playground: fix and extend code-in-URL handling (#1434)

* percent-encode base-64 code when sharing
* to fix old share links, convert ' ' to '+' before base-64 decoding
* add #rawcode= as alternative to #code= (takes plain non-base64, but percent-encoded text)

AshineFoster committed:
Allow ocaml.org to run while offline (#1366)

When offline, ocaml.org uses the current local state of opam-repository. This enables working on the codebase even when there is no Internet access.

tmattio committed:
Partially import the OCaml Platform roadmap (#1432)

sabine committed:
Revert "Partially import the OCaml Platform roadmap (#1432)" (#1464)

This reverts commit 89ff64a58962bfca9cef17199cdeb20871ae526e.

github-actions[bot] committed:
[create-pull-request] automated change (#1468)

SaySayo committed:
Skill Tag Component [Learn Area Redesign] (#1427)

github-actions[bot] committed:
[create-pull-request] automated change (#1472)

Zineb-Ada committed:
Merge branch 'blogs_filter' of github.com:Zineb-Ada/ocaml.org into blogs_filter

Zineb-Ada committed:
Change the tags of the Opam blogs

tmattio commented:
I understand that @Zineb-Ada won't be working on this anymore, and this doesn't seem to be in a state to be merged quickly, so I'm closing this, but if anyone would like to continue this work, a new PR will be welcome.
